legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java

```diff
@@ -428,4 +451,33 @@ private Optional<Long> getNextBatchId(Datasets datasets, Executor<SqlGen, Tabula
         }
         return Optional.empty();
     }
+
+    private Optional<Map<OptimizationFilter, List<Object>>> getOptimizationFilterBounds(Datasets datasets, Executor<SqlGen, TabularData, SqlPlan> executor,
+                                          Transformer<SqlGen, SqlPlan> transformer, IngestMode ingestMode)
+    {
+        List<OptimizationFilter> filters = ingestMode.accept(IngestModeVisitors.RETRIEVE_OPTIMIZATION_FILTERS);
+        if (!filters.isEmpty())
+        {
+            Map<OptimizationFilter, List<Object>> map = new HashMap<>();
+            for (OptimizationFilter filter : filters)
+            {
+                LogicalPlan logicalPlanForMinAndMaxForField = LogicalPlanFactory.getLogicalPlanForMinAndMaxForField(datasets.stagingDataset(), filter.fieldName());
+                List<TabularData> tabularData = executor.executePhysicalPlanAndGetResults(transformer.generatePhysicalPlan(logicalPlanForMinAndMaxForField));
+                Map<String, Object> resultMap = tabularData.stream()
+                    .findFirst()
+                    .map(TabularData::getData)
+                    .flatMap(t -> t.stream().findFirst())
+                    .orElseThrow(IllegalStateException::new);
+                // Put into map only when not null
+                Object lower = resultMap.get(MIN_OF_FIELD);
+                Object upper = resultMap.get(MAX_OF_FIELD);
+                if (lower != null && upper != null)
+                {
+                    map.put(filter, Arrays.asList(lower, upper));
```

Comment:
Contract will be cleaner if we put this on a Tuple, as we always will have two values rather than a list.  On the caller, it will be tuple.first, tuples.second rather than 'magic' index values. for the array.

Reply:
Sure, fixed this!